### PR TITLE
Port to Django

### LINKED ISSUE #49: Drop Server does not export Prometheus Metrics
Even while listening on the port Configured for Prometheus Metrics it does not export Metrics.
```
 curl localhost:8501/metrics -v
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8501 (#0)
> GET /metrics HTTP/1.1
> User-Agent: curl/7.35.0
> Host: localhost:8501
> Accept: */*
> 
* Recv failure: Connection reset by peer
* Closing connection 0
curl: (56) Recv failure: Connection reset by peer
```

### LINKED ISSUE #52: Port to Django from Flask
We don't need two different frameworks filling the same niche in our MSA.

Tests look good so porting isn't ~~expected~~ estimated to be a large effort. 1-2 points on the serverpoint-scale.

### PR DESCRIPTION
**Pull request by enkore**
Known differences:

- If an invalid drop id is passed in a POST request the same error
  a GET request would yield is returned, instead of a blank body.
- database error handling: 51b8f6c
- Management and configuration is different ofc.
- Prometheus stats are exported like on qabel-accounting, via /metrics
  - And they work (Fixes #49)

Other:

- Recycled naming scheme from ~~accounting~~ register: qabel_xxx = project name, xy_service = app name
- Database schema *should* be compatible:

    Old code with SQLAlchemy (sqlite3 .TABLE)

        CREATE TABLE drops (
                id INTEGER NOT NULL, 
                drop_id VARCHAR, 
                message BLOB, 
                created_at DATETIME, 
                PRIMARY KEY (id)
        );

    Django schema

        CREATE TABLE "drops" (
           "id" integer NOT NULL PRIMARY KEY AUTOINCREMENT,
           "drop_id" varchar(43) NOT NULL,
           "message" BLOB NOT NULL,
           "created_at" datetime NOT NULL
        );

    The additional NOT NULLs are imho covered by the verification done:

      - created_at can't ever be null
      - empty messages aren't allowed
      - drop_id is always precisely 43 characters
- Fixes #52 

TODO:

- [ ] Docs update
- [x] remove admin
- [x] remove REQUEST_TIME monitoring
- [x] separate get and head views
- [x] fixed build.xml for CI

### CODE REVIEW

**Comment by enkore** [comment]
This is now also in the testing inv scripts, but didn't really test. ;)

File: drop_service/monitoring.py [hunk 1]
```diff
@@ -1,23 +1,11 @@
 from prometheus_client import Summary, Counter
-import threading
 
+# Remove this? It's a bit redundant with django-prometheus around which generates very similar stats (among other things).
```
Comment (audax) [inline drop_service/monitoring.py]: yes, remove it.

File: drop_service/views.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            return HttpResponse(status=status.HTTP_204_NO_CONTENT)
+
+        have_since, since = self.get_if_modified_since()
+        if have_since:
+            drops = drops.filter(created_at__gt=since)
+            if not drops:
+                log(status.HTTP_304_NOT_MODIFIED)
+                return HttpResponseNotModified()
+
+        if request.method == 'GET':
```
Comment (audax) [inline drop_service/views.py]: Is this needed? Are head requests routed to get() ?

**Comment by audax** [comment]
Update the readme and I'll merge it

File: drop_service/views.py (hunk 2 above)
Comment (enkore) [inline drop_service/views.py]: Yes, HEAD requests are routed to the get function by default.

> By default, a HEAD request will be delegated to get(). If you need to handle HEAD requests in a different way than GET, you can override the head() method.
> https://docs.djangoproject.com/ja/1.9/ref/class-based-views/base/#django.views.generic.base.View.dispatch

I think I'll just split this into get() and head(), that'd be cleaner.

**Comment by enkore** [comment]
By the way, if we skip the admin we can literally remove half of the middleware and disable a lot of code. I'm thinking - add a feature flag for that?

**Comment by audax** [comment]
I don't think we need the admin interface in the drop server, kill it.

File: drop_service/views.py (hunk 2 above)
Comment (audax) [inline drop_service/views.py]: Eggcelent. :egg:

**Comment by audax** [comment]
```
13:48:26.676      [exec] ============================= test session starts ==============================
13:48:26.676      [exec] platform linux -- Python 3.5.1+, pytest-2.8.3, py-1.4.31, pluggy-0.3.1 -- /var/lib/go-agent/pipelines/drop-build/venv/bin/python3.5
13:48:26.676      [exec] cachedir: ../.cache
13:48:26.676      [exec] django settings: qabel_drop.settings (from ini file)
13:48:26.676      [exec] run-last-failure: run all (no recorded failures)
13:48:26.676      [exec] rootdir: /var/lib/go-agent/pipelines/drop-build, inifile: pytest.ini
13:48:26.676      [exec] plugins: django-2.9.1
13:48:26.676      [exec] collecting ... collected 25 items
13:48:26.676      [exec] 
13:48:27.076      [exec] test_monitoring.py::test_ok FAILEDDestroying test database for alias 'default'...
13:48:27.076      [exec] 
13:48:27.076      [exec] 
13:48:27.077      [exec] =================================== FAILURES ===================================
13:48:27.077      [exec] ___________________________________ test_ok ____________________________________
13:48:27.077      [exec] 
13:48:27.077      [exec] client = <django.test.client.Client object at 0x7fd83e8c7588>
13:48:27.077      [exec] drop_messages = (<abcdefghijklmnopqrstuvwxyzabcdefghijklmnopo for b'Hello World' created at 2016-07-04 13:48:26.688078+00:00 >, <abcdefghijklmnopqrstuvwxyzabcdefghijklmnfoo for b'Bar' created at 2016-01-01 00:00:00+00:00 >)
13:48:27.077      [exec] registry = <class 'conftest.Registry'>
13:48:27.077      [exec] 
13:48:27.077      [exec]     def test_ok(client, drop_messages, registry):
13:48:27.077      [exec]         with registry.assert_drop_sent(count=1):
13:48:27.077      [exec] >           response = client.get('/abcdefghijklmnopqrstuvwxyzabcdefghijklmnopo')
13:48:27.077      [exec] 
13:48:27.077      [exec] test_monitoring.py:10: 
13:48:27.077      [exec] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
13:48:27.077      [exec] ../venv/lib/python3.5/site-packages/django/test/client.py:503: in get
13:48:27.077      [exec]     **extra)
13:48:27.077      [exec] ../venv/lib/python3.5/site-packages/django/test/client.py:304: in get
13:48:27.078      [exec]     return self.generic('GET', path, secure=secure, **r)
13:48:27.078      [exec] ../venv/lib/python3.5/site-packages/django/test/client.py:380: in generic
13:48:27.078      [exec]     return self.request(**r)
13:48:27.078      [exec] ../venv/lib/python3.5/site-packages/django/test/client.py:467: in request
13:48:27.078      [exec]     six.reraise(*exc_info)
13:48:27.078      [exec] ../venv/lib/python3.5/site-packages/django/utils/six.py:686: in reraise
13:48:27.078      [exec]     raise value
13:48:27.078      [exec] ../venv/lib/python3.5/site-packages/django/core/handlers/base.py:149: in get_response
13:48:27.078      [exec]     response = self.process_exception_by_middleware(e, request)
13:48:27.078      [exec] ../venv/lib/python3.5/site-packages/django/core/handlers/base.py:147: in get_response
13:48:27.078      [exec]     response = wrapped_callback(request, *callback_args, **callback_kwargs)
13:48:27.078      [exec] ../venv/lib/python3.5/site-packages/django/views/generic/base.py:68: in view
13:48:27.078      [exec]     return self.dispatch(request, *args, **kwargs)
13:48:27.078      [exec] ../venv/lib/python3.5/site-packages/django/utils/decorators.py:67: in _wrapper
13:48:27.078      [exec]     return bound_func(*args, **kwargs)
13:48:27.079      [exec] ../venv/lib/python3.5/site-packages/django/views/decorators/csrf.py:58: in wrapped_view
13:48:27.079      [exec]     return view_func(*args, **kwargs)
13:48:27.079      [exec] ../venv/lib/python3.5/site-packages/django/utils/decorators.py:63: in bound_func
13:48:27.079      [exec]     return func.__get__(self, type(self))(*args2, **kwargs2)
13:48:27.079      [exec] ../drop_service/views.py:31: in dispatch
13:48:27.079      [exec]     return super().dispatch(request, *args, **kwargs)
13:48:27.079      [exec] ../venv/lib/python3.5/site-packages/django/views/generic/base.py:88: in dispatch
13:48:27.079      [exec]     return handler(request, *args, **kwargs)
13:48:27.079      [exec] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
13:48:27.079      [exec] 
13:48:27.079      [exec] self = <drop_service.views.DropView object at 0x7fd83e7c6d68>
13:48:27.079      [exec] request = <WSGIRequest: GET '/abcdefghijklmnopqrstuvwxyzabcdefghijklmnopo'>
13:48:27.079      [exec] drop_id = 'abcdefghijklmnopqrstuvwxyzabcdefghijklmnopo'
13:48:27.079      [exec] 
13:48:27.079      [exec]     def get(self, request, drop_id):
13:48:27.079      [exec] >       response, drops = self._get_drops(request, drop_id)
13:48:27.080      [exec] E       TypeError: _get_drops() takes 2 positional arguments but 3 were given
13:48:27.080      [exec] 
13:48:27.080      [exec] ../drop_service/views.py:51: TypeError
13:48:27.080      [exec] ---------------------------- Captured stdout setup -----------------------------
13:48:27.080      [exec] Creating test database for alias 'default'...
13:48:27.080      [exec] ----- generated xml file: /var/lib/go-agent/pipelines/drop-build/junit.xml -----
13:48:27.080      [exec] !!!!!!!!!!!!!!!!!!!! Interrupted: stopping after 1 failures !!!!!!!!!!!!!!!!!!!!
13:48:27.080      [exec] =========================== 1 failed in 0.44 seconds ===========================
```

We really need to include the drop server in the CI.

**Comment by julianseeger** [comment]
drop-PR is now included in the CI-Server, so @enkore place a new PR and it will be built

**Comment by enkore** [comment]
Sorry, for the fuckup. Must have removed that unused parameter without retesting before committing.

**Pull request merged**